File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/data/api/factories/OkHttpClientFactory.java
package com.crpr.androidcinema.data.api.factories;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;

/**
 * Created by cribeiro on 08/01/2016.
 */
public class OkHttpClientFactory {

    public static final int CONFIG_CLIENT = 1;

    private final InterceptorsFactory _factory;

    public OkHttpClientFactory(InterceptorsFactory factory){
        _factory = factory;
    }

    public OkHttpClient getClient(int type){
        OkHttpClient.Builder builder = new OkHttpClient().newBuilder();

        builder.connectTimeout(30, TimeUnit.SECONDS);
        builder.readTimeout(30, TimeUnit.SECONDS);

        switch (type){
            case CONFIG_CLIENT:
                builder.addNetworkInterceptor(_factory.get(InterceptorsFactory.LOG));
                builder.addNetworkInterceptor(_factory.get(InterceptorsFactory.AUTH));
                builder.addInterceptor(_factory.get(InterceptorsFactory.HTTP));
                break;

            default:
                builder.addNetworkInterceptor(_factory.get(InterceptorsFactory.LOG));
                builder.addInterceptor(_factory.get(InterceptorsFactory.HTTP));
                break;
        }

        return builder.build();
    }
}

File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/data/api/factories/OkHttpClientFactory.kt
package com.crpr.androidcinema.data.api.factories

import okhttp3.OkHttpClient
import java.util.concurrent.TimeUnit

/**
 * Created by cribeiro on 08/01/2016.
 */
class OkHttpClientFactory(private val _factory: InterceptorsFactory) {

    fun getClient(type: Int): OkHttpClient {
        val builder = OkHttpClient().newBuilder()

        builder.connectTimeout(30, TimeUnit.SECONDS)
        builder.readTimeout(30, TimeUnit.SECONDS)

        when (type) {
            STANDARD_CLIENT -> {
                builder.addNetworkInterceptor(_factory[InterceptorsFactory.LOG])
                builder.addNetworkInterceptor(_factory[InterceptorsFactory.AUTH])
                builder.addInterceptor(_factory[InterceptorsFactory.HTTP])
            }

            else -> {
                builder.addNetworkInterceptor(_factory[InterceptorsFactory.LOG])
                builder.addInterceptor(_factory[InterceptorsFactory.HTTP])
            }
        }

        return builder.build()
    }

    companion object {

        val STANDARD_CLIENT = 1
    }
}
